File: common/src/androidMain/kotlin/com/blockstream/common/devices/APDUResponse.java
//package com.blockstream.common.devices;
//
//import java.io.*;
//
///**
// * ISO7816-4 APDU response.
// */
//public class APDUResponse {
//  public static final int SW_OK = 0x9000;
//  public static final int SW_SECURITY_CONDITION_NOT_SATISFIED = 0x6982;
//  public static final int SW_AUTHENTICATION_METHOD_BLOCKED = 0x6983;
//  public static final int SW_CARD_LOCKED = 0x6283;
//  public static final int SW_REFERENCED_DATA_NOT_FOUND = 0x6A88;
//  public static final int SW_CONDITIONS_OF_USE_NOT_SATISFIED = 0x6985; // applet may be already installed
//  public static final int SW_WRONG_PIN_MASK = 0x63C0;
//  public static final int SW_WRONG_PIN_LEGACY = 0x9C02;
//  public static final int SW_BLOCKED_PIN = 0x9C0C;
//  public static final int SW_FACTORY_RESET = 0xFF00;
//  public static final String HEXES = "0123456789ABCDEF";
//
//  private byte[] apdu;
//  private byte[] data;
//  private int sw;
//  private int sw1;
//  private int sw2;
//
//  /**
//   * Creates an APDU object by parsing the raw response from the card.
//   *
//   * @param apdu the raw response from the card.
//   */
//  public APDUResponse(byte[] apdu)  {
//    if (apdu.length < 2) {
//      throw new IllegalArgumentException("APDU response must be at least 2 bytes");
//    }
//    this.apdu = apdu;
//    this.parse();
//  }
//
//  public APDUResponse(byte[] data, byte sw1, byte sw2)  {
//    byte[] apdu= new byte[data.length + 2];
//    System.arraycopy(data, 0, apdu, 0, data.length);
//    apdu[data.length]= sw1;
//    apdu[data.length+1]= sw2;
//    this.apdu = apdu;
//    this.parse();
//  }
//
//
//  /**
//   * Parses the APDU response, separating the response data from SW.
//   */
//  private void parse() {
//    int length = this.apdu.length;
//
//    this.sw1 = this.apdu[length - 2] & 0xff;
//    this.sw2 = this.apdu[length - 1] & 0xff;
//    this.sw = (this.sw1 << 8) | this.sw2;
//
//    this.data = new byte[length - 2];
//    System.arraycopy(this.apdu, 0, this.data, 0, length - 2);
//  }
//
//  /**
//   * Returns true if the SW is 0x9000.
//   *
//   * @return true if the SW is 0x9000.
//   */
//  public boolean isOK() {
//    return this.sw == SW_OK;
//  }
//
//  /**
//   * Asserts that the SW is 0x9000. Throws an exception if it isn't
//   *
//   * @return this object, to simplify chaining
//   * @throws APDUException if the SW is not 0x9000
//   */
//  public APDUResponse checkOK() throws APDUException {
//    return this.checkSW(SW_OK);
//  }
//
//  /**
//   * Asserts that the SW is contained in the given list. Throws an exception if it isn't.
//   *
//   * @param codes the list of SWs to match.
//   * @return this object, to simplify chaining
//   * @throws APDUException if the SW is not 0x9000
//   */
//  public APDUResponse checkSW(int... codes) throws APDUException {
//    for (int code : codes) {
//      if (this.sw == code) {
//        return this;
//      }
//    }
//
//    switch (this.sw) {
//      case SW_SECURITY_CONDITION_NOT_SATISFIED:
//        throw new APDUException(this.sw, "security condition not satisfied");
//      case SW_AUTHENTICATION_METHOD_BLOCKED:
//        throw new APDUException(this.sw, "authentication method blocked");
//      default:
//        throw new APDUException(this.sw,  "Unexpected error SW");
//    }
//  }
//
//  /**
//   * Asserts that the SW is 0x9000. Throws an exception with the given message if it isn't
//   *
//   * @param message the error message
//   * @return this object, to simplify chaining
//   * @throws APDUException if the SW is not 0x9000
//   */
//  public APDUResponse checkOK(String message) throws APDUException {
//    return checkSW(message, SW_OK);
//  }
//
//  /**
//   * Asserts that the SW is contained in the given list. Throws an exception with the given message if it isn't.
//   *
//   * @param message the error message
//   * @param codes the list of SWs to match.
//   * @return this object, to simplify chaining
//   * @throws APDUException if the SW is not 0x9000
//   */
//  public APDUResponse checkSW(String message, int... codes) throws APDUException {
//    for (int code : codes) {
//      if (this.sw == code) {
//        return this;
//      }
//    }
//
//    throw new APDUException(this.sw, message);
//  }
//
//  /**
//   * Checks response from an authentication command (VERIFY PIN, UNBLOCK PUK)
//   *
//   * @throws WrongPINException wrong PIN
//   * @throws APDUException unexpected response
//   */
////  public APDUResponse checkAuthOK() throws WrongPINException, WrongPINLegacyException, BlockedPINException, APDUException {
////    if ((this.sw & SW_WRONG_PIN_MASK) == SW_WRONG_PIN_MASK) {
////      throw new WrongPINException(sw2 & 0x0F);
////    } else if (this.sw == SW_WRONG_PIN_LEGACY) {
////      throw new WrongPINLegacyException();
////    } else if (this.sw == SW_BLOCKED_PIN) {
////      throw new BlockedPINException();
////    } else if (this.sw == SW_FACTORY_RESET) {
////      throw new ResetToFactoryException();
////    } else {
////      return checkOK();
////    }
////  }
//
//  /**
//   * Returns the data field of this APDU.
//   *
//   * @return the data field of this APDU
//   */
//  public byte[] getData() {
//    return this.data;
//  }
//
//  /**
//   * Returns the Status Word.
//   *
//   * @return the status word
//   */
//  public int getSw() {
//    return this.sw;
//  }
//
//  /**
//   * Returns the SW1 byte
//   * @return SW1
//   */
//  public int getSw1() {
//    return this.sw1;
//  }
//
//  /**
//   * Returns the SW2 byte
//   * @return SW2
//   */
//  public int getSw2() {
//    return this.sw2;
//  }
//
//  /**
//   * Returns the raw unparsed response.
//   *
//   * @return raw APDU data
//   */
//  public byte[] getBytes() {
//    return this.apdu;
//  }
//
//  /**
//   * Serializes the APDU to human readable hex string format
//   *
//   * @return the hex string representation of the APDU
//   */
//  public String toHexString() {
//		byte[] raw= this.apdu;
//		try{
//      if ( raw == null ) {
//             return "";
//          }
//          final StringBuilder hex = new StringBuilder( 2 * raw.length );
//          for ( final byte b : raw ) {
//             hex.append(HEXES.charAt((b & 0xF0) >> 4))
//           .append(HEXES.charAt((b & 0x0F)));
//          }
//          return hex.toString();
//    } catch(Exception e){
//      return "Exception in APDUResponse.toHexString()";
//    }
//  }
//}

File: common/src/androidMain/kotlin/com/blockstream/common/devices/ApduResponse.kt
package com.blockstream.common.devices

/**
 * ISO7816-4 APDU response.
 */
class ApduResponse {
    private var apdu: ByteArray = byteArrayOf()
    private var data: ByteArray = byteArrayOf()
    private var sw: Int = 0
    private var sw1: Int = 0
    private var sw2: Int = 0

    /**
     * Creates an APDU object by parsing the raw response from the card.
     *
     * @param apdu the raw response from the card.
     */
    constructor(apdu: ByteArray) {
        require(apdu.size >= 2) { "APDU response must be at least 2 bytes" }
        this.apdu = apdu
        parse()
    }

    constructor(data: ByteArray, sw1: Byte, sw2: Byte) {
        val apduArray = ByteArray(data.size + 2)
        System.arraycopy(data, 0, apduArray, 0, data.size)
        apduArray[data.size] = sw1
        apduArray[data.size + 1] = sw2
        this.apdu = apduArray
        parse()
    }

    /**
     * Parses the APDU response, separating the response data from SW.
     */
    private fun parse() {
        val length = apdu.size

        sw1 = apdu[length - 2].toInt() and 0xff
        sw2 = apdu[length - 1].toInt() and 0xff
        sw = (sw1 shl 8) or sw2

        data = ByteArray(length - 2)
        System.arraycopy(apdu, 0, data, 0, length - 2)
    }

    /**
     * Returns true if the SW is 0x9000.
     *
     * @return true if the SW is 0x9000.
     */
    fun isOK(): Boolean = sw == SW_OK

    /**
     * Asserts that the SW is 0x9000. Throws an exception if it isn't
     *
     * @return this object, to simplify chaining
     * @throws ApduException if the SW is not 0x9000
     */
    @Throws(ApduException::class)
    fun checkOK(): ApduResponse = checkSW(SW_OK)

    /**
     * Asserts that the SW is contained in the given list. Throws an exception if it isn't.
     *
     * @param codes the list of SWs to match.
     * @return this object, to simplify chaining
     * @throws ApduException if the SW is not 0x9000
     */
    @Throws(ApduException::class)
    fun checkSW(vararg codes: Int): ApduResponse {
        for (code in codes) {
            if (sw == code) {
                return this
            }
        }

        when (sw) {
            SW_SECURITY_CONDITION_NOT_SATISFIED ->
                throw ApduException(sw, "security condition not satisfied")
            SW_AUTHENTICATION_METHOD_BLOCKED ->
                throw ApduException(sw, "authentication method blocked")
            else ->
                throw ApduException(sw, "Unexpected error SW")
        }
    }

    /**
     * Asserts that the SW is 0x9000. Throws an exception with the given message if it isn't
     *
     * @param message the error message
     * @return this object, to simplify chaining
     * @throws ApduException if the SW is not 0x9000
     */
    @Throws(ApduException::class)
    fun checkOK(message: String): ApduResponse = checkSW(message, SW_OK)

    /**
     * Asserts that the SW is contained in the given list. Throws an exception with the given message if it isn't.
     *
     * @param message the error message
     * @param codes the list of SWs to match.
     * @return this object, to simplify chaining
     * @throws ApduException if the SW is not 0x9000
     */
    @Throws(ApduException::class)
    fun checkSW(message: String, vararg codes: Int): ApduResponse {
        for (code in codes) {
            if (sw == code) {
                return this
            }
        }
        throw ApduException(sw, message)
    }

    /**
     * Serializes the APDU to human readable hex string format
     *
     * @return the hex string representation of the APDU
     */
    fun toHexString(): String {
        return try {
            if (apdu.isEmpty()) {
                ""
            } else {
                StringBuilder(2 * apdu.size).apply {
                    apdu.forEach { b ->
                        append(HEXES[(b.toInt() and 0xF0) shr 4])
                        append(HEXES[b.toInt() and 0x0F])
                    }
                }.toString()
            }
        } catch (e: Exception) {
            "Exception in ApduResponse.toHexString()"
        }
    }

    companion object {
        const val SW_OK = 0x9000
        const val SW_SECURITY_CONDITION_NOT_SATISFIED = 0x6982
        const val SW_AUTHENTICATION_METHOD_BLOCKED = 0x6983
        const val SW_CARD_LOCKED = 0x6283
        const val SW_REFERENCED_DATA_NOT_FOUND = 0x6A88
        const val SW_CONDITIONS_OF_USE_NOT_SATISFIED = 0x6985 // applet may be already installed
        const val SW_WRONG_PIN_MASK = 0x63C0
        const val SW_WRONG_PIN_LEGACY = 0x9C02
        const val SW_BLOCKED_PIN = 0x9C0C
        const val SW_FACTORY_RESET = 0xFF00
        const val HEXES = "0123456789ABCDEF"
    }

    // Getters converted to Kotlin properties
    fun getData(): ByteArray = data
    fun getSw(): Int = sw
    fun getSw1(): Int = sw1
    fun getSw2(): Int = sw2
    fun getBytes(): ByteArray = apdu
}